spec_validator.go

```diff
@@ -0,0 +1,147 @@
+	"strconv"
+	"strings"
+	"unicode"
+
+	"github.com/google/uuid"
+)
+
+const (
+	serialPrefix = "serial"
+	uuidPrefix   = "uuid"
+	eventPrefix  = "event"
+	dnsPrefix    = "dns"
+)
+
+var (
+	ErrorInvalidMessageEncoding = errors.New("invalid message encoding")
+	ErrorInvalidMessageType     = errors.New("invalid message type")
+	ErrorInvalidSource          = errors.New("invalid Source name")
+	ErrorInvalidDestination     = errors.New("invalid Destination name")
+)
+
+// SpecValidators is a WRP validator that ensures messages are valid based on
+// each spec validator in the list. Only validates the opinionated portions of the spec.
+// SpecValidators validates the following fields: UTF8 (all string fields), MessageType, Source, Destination
+func SpecValidators() Validators {
+	return Validators{UTF8Validator, MessageTypeValidator, SourceValidator, DestinationValidator}
+}
+
+// UTF8Validator is a WRP validator that takes messages and validates that it contains UTF-8 strings.
+var UTF8Validator ValidatorFunc = func(m Message) error {
+	if err := UTF8(m); err != nil {
+		return fmt.Errorf("%w: %v", ErrorInvalidMessageEncoding, err)
+	}
+
+	return nil
+}
+
+// MessageTypeValidator is a WRP validator that takes messages and validates their Type.
+var MessageTypeValidator ValidatorFunc = func(m Message) error {
+	if m.Type < Invalid0MessageType || m.Type > lastMessageType {
+		return ErrorInvalidMessageType
+	}
+
+	switch m.Type {
+	case Invalid0MessageType, Invalid1MessageType, lastMessageType:
+		return ErrorInvalidMessageType
+	}
+
+	return nil
+}
+
+// SourceValidator is a WRP validator that takes messages and validates their Source.
+// Only mac and uuid sources are validated. Serial, event and dns sources are
+// not validated.
+var SourceValidator ValidatorFunc = func(m Message) error {
+	if err := validateLocator(m.Source); err != nil {
+		return fmt.Errorf("%w: %v", ErrorInvalidSource, err)
+	}
+
+	return nil
+}
+
+// DestinationValidator is a WRP validator that takes messages and validates their Destination.
+// Only mac and uuid destinations are validated. Serial, event and dns destinations are
+// not validated.
+var DestinationValidator ValidatorFunc = func(m Message) error {
+	if err := validateLocator(m.Destination); err != nil {
+		return fmt.Errorf("%w: %v", ErrorInvalidDestination, err)
+	}
+
+	return nil
+}
+
+// locatorPattern is the precompiled regular expression that all source and dest locators must match.
+// Matching is partial, as everything after the authority (ID) is ignored. https://xmidt.io/docs/wrp/basics/#locators
+var locatorPattern = regexp.MustCompile(
+	`^(?P<scheme>(?i)` + macPrefix + `|` + uuidPrefix + `|` + eventPrefix + `|` + dnsPrefix + `|` + serialPrefix + `):(?P<authority>[^/]+)?`,
+)
+
+// validateLocator validates a given locator's scheme and authority (ID).
+// Only mac and uuid schemes' IDs are validated. IDs from serial, event and dns schemes are
+// not validated.
+func validateLocator(l string) error {
```

Comment:
I would make all of the errors in this function have some unexported error vars that we can check against in the tests.  Then the code kind of documents itself, and testing is easy.  In places where we're returning a constant string error currently, the var can be returned instead without any `fmt.Errorf` calls in the function.